bsync: init at 20231221-25f7773

dietmarw opened the pull request:
## Description of changes
Initial version for nixpkgs

Bsync is a bidirectional file synchronization tool, using rsync for transfers. Moved files are also synchronized in a smart way.

https://github.com/dooblem/bsync

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

bjornfor commented on pkgs/applications/networking/sync/bsync/default.nix:
```diff
@@ -0,0 +1,31 @@
+    homepage = "https://github.com/dooblem/bsync";
+    description = "Bidirectional Synchronization using Rsync";
+    license = licenses.gpl3Plus;
+    maintainers = with maintainers;  [ "dietmarw" ];
```
You can rebase on master and fix this line so it references your new maintainers entry.

bjornfor commented on pkgs/applications/networking/sync/bsync/default.nix:
```diff
@@ -0,0 +1,31 @@
+    maintainers = with maintainers;  [ "dietmarw" ];
+    platforms = with lib.platforms; linux ++ darwin;
+  };
+
```
There is usually no blank line here. Please remove.

dietmarw replied:
Removed in 71dc388339b95a26e4d64745fd77287999fd9dc1

bjornfor commented:
New packages go to pkgs/by-name/...

h7x4 commented on pkgs/applications/networking/sync/bsync/default.nix:
```diff
@@ -0,0 +1,31 @@
+    hash   = "sha256-k25MjLis0/dp1TTS4aFeJZq/c0T01LmNcWtC+dw/kKY=";
+  };
+
+  installPhase = ''
+    mkdir -p $out/bin
+    cp bsync $out/bin
+    chmod +x $out/bin/bsync
+  '';
```
```suggestion
  installPhase = ''
    runHook preInstall
    install -Dm555 bsync -t $out/bin
    runHook postInstall
  '';
```

h7x4 commented on pkgs/applications/networking/sync/bsync/default.nix:
```diff
@@ -0,0 +1,31 @@
+, stdenv
+}:
+
+stdenv.mkDerivation rec {
```
I don't think the `rec` is necessary here

```suggestion
stdenv.mkDerivation {
```

h7x4 commented on pkgs/applications/networking/sync/bsync/default.nix:
```diff
@@ -0,0 +1,31 @@
+    description = "Bidirectional Synchronization using Rsync";
+    license = licenses.gpl3Plus;
+    maintainers = with maintainers;  [ "dietmarw" ];
+    platforms = with lib.platforms; linux ++ darwin;
```
I assume this is the main executable?

Also I think it's safe to assume it would run on the other unix-based systems as well.

```suggestion
    platforms = platforms.unix;
    mainProgram = "bsync";
```

dietmarw commented:
@bjornfor and @h7x4 Thanks for the feedback. Very much appreciated especially since this is my first package and I'm still in the learning phase how things are done properly. 
Now I have rebased the PR as suggested and fixed all the other issues. I guess I should really squash my commits to just one. But this means also the reference for the review is gone then (though this is probably the case any way after the rebase). 

Let me know if I should do a squash or not.

h7x4 commented:
Please squash the commits :)

If you'd like to keep the commits as reference (although I suggest you could just use the now resolved github comments as well), you could checkout a new branch, and switch back to `init-bsync` before squashing. That way, the commits will stay as they are on the new branch.

h7x4 commented on pkgs/by-name/bs/bsync/package.nix:
```diff
@@ -0,0 +1,31 @@
+
+stdenv.mkDerivation {
+  pname = "bsync";
+  version = "20231221-25f7773";
```
I think unstable versions are usually done in this format. You'll have to redo the commit message as well

```suggestion
  version = "unstable-2023-12-21";
```

h7x4 commented on pkgs/by-name/bs/bsync/package.nix:
```diff
@@ -0,0 +1,31 @@
+    homepage = "https://github.com/dooblem/bsync";
+    description = "Bidirectional Synchronization using Rsync";
+    license = licenses.gpl3Plus;
+    maintainers = with maintainers;  [ "dietmarw" ];
```
Assuming you've rebased this branch on top of a version of master that contains the commit where you add yourself as a maintainer, this line should look like this:

```suggestion
    maintainers = with maintainers;  [ dietmarw ];
```

Feel free to ask if you're not used with this kind of git usage.

dietmarw commented:
Thanks again. I fixed the version to follow the convention and have also changed the commit message (thanks to GitKraken such operation are just a breeze ;-) )

h7x4 commented on pkgs/by-name/bs/bsync/package.nix:
```diff
@@ -0,0 +1,31 @@
+    install -Dm555 bsync -t $out/bin
+    runHook postInstall
+  '';
+
```
I had a look at upstream. The program seems to have a bunch of runtime dependencies, that we'll need to add to the `$PATH` of the program. These ones were the ones I found, but there might be something I've overlooked. You'll have to add the packages referenced to the arguments up top as well.

```suggestion

nativeBuildInputs = [ makeWrapper ];
buildInputs = [ python3 ];

fixupPhase = ''
  runHook preFixup

  patchShebangs $out/bin/bsync
  wrapProgram $out/bin/bsync \
    --prefix PATH ":" ${lib.makeLibraryPath [ openssh rsync findutils ]}

  runHook postFixup
'';

```

h7x4 replied:
Yes, I think you should be able open a pure nix-shell with the package like this:

```
nix-shell --pure -A bsync
```

dietmarw commented:
I tried 
```shell
$ nix-shell --pure -A result/bin/bsync 
error: getting status of '/home/dietmarw/.workspace/nixpkgs/pkgs/by-name/bs/bsync/default.nix': No such file or directory
```
but that is probably not the way to test this.

h7x4 commented:
Right, `-A` is supposed to point at the attribute in `default.nix`, not the file itself. I think `nix-shell --pure -A bsync` should work?

h7x4 commented on pkgs/by-name/bs/bsync/package.nix:
```diff
@@ -0,0 +1,49 @@
+, python3
+, openssh
+, rsync
+, findutils
+}:
+
+stdenv.mkDerivation {
+  pname = "bsync";
+  version = "unstable-2023-12-21";
+
+  src = fetchFromGitHub {
+    owner  = "dooblem";
+    repo   = "bsync";
+    rev    = "25f77730750720ad68b0ab2773e79d9ca98c3647";
+    hash   = "sha256-k25MjLis0/dp1TTS4aFeJZq/c0T01LmNcWtC+dw/kKY=";
+  };
+
+  installPhase = ''
+    runHook preInstall
+    install -Dm555 bsync -t $out/bin
+    runHook postInstall
+  '';
+
+nativeBuildInputs = [ makeWrapper ];
+buildInputs = [ python3 ];
+
+fixupPhase = ''
+  runHook preFixup
+
+  patchShebangs $out/bin/bsync
+  wrapProgram $out/bin/bsync \
+    --prefix PATH ":" ${lib.makeLibraryPath [ openssh rsync findutils ]}
```
Seems like the program wants `which` as well: https://github.com/dooblem/bsync/blob/25f77730750720ad68b0ab2773e79d9ca98c3647/bsync#L158

```suggestion
, findutils
, which
}:

stdenv.mkDerivation {
  pname = "bsync";
  version = "unstable-2023-12-21";

  src = fetchFromGitHub {
    owner  = "dooblem";
    repo   = "bsync";
    rev    = "25f77730750720ad68b0ab2773e79d9ca98c3647";
    hash   = "sha256-k25MjLis0/dp1TTS4aFeJZq/c0T01LmNcWtC+dw/kKY=";
  };

  installPhase = ''
    runHook preInstall
    install -Dm555 bsync -t $out/bin
    runHook postInstall
  '';

nativeBuildInputs = [ makeWrapper ];
buildInputs = [ python3 ];

fixupPhase = ''
  runHook preFixup

  patchShebangs $out/bin/bsync
  wrapProgram $out/bin/bsync \
    --prefix PATH ":" ${lib.makeLibraryPath [ openssh rsync findutils which ]}
```

dietmarw commented:
> Right, `-A` is supposed to point at the attribute in `default.nix`, not the file itself. I think `nix-shell --pure -A bsync` should work?

Not really since there is no bsync installed on my system. Only the one I'm building. On another machine, I include that local package as part of my home-config. But I don't have access to that one right now.

h7x4 commented on pkgs/by-name/bs/bsync/package.nix:
```diff
@@ -0,0 +1,50 @@
+  meta = with lib; {
+    homepage = "https://github.com/dooblem/bsync";
+    description = "Bidirectional Synchronization using Rsync";
+    license = licenses.gpl3Plus;
```
The license upstream seems like the normal GPL 3

```suggestion
    license = licenses.gpl3Only;
```

dietmarw replied:
Thanks for spotting this.

h7x4 commented:
I think this is looking really good. I see that there are some tests upstream as well, but it would require some shuffling of phases and refactoring, so it might be okay to do that later if needed.

dietmarw committed:
bsync: init at unstable-2023-12-21

No release so using unstable-DATE convention for version

h7x4 commented:
Okay, just a rebase left

dietmarw commented:
@h7x4 Thanks for your thorough review and feedback. Yes the tests is definitely something that could be added. I'll try to tackle that hopefully later when I'm a bit more proficient with Nix. This PR and the reviews taught me already quite a lot today :-)

dietmarw commented:
> Okay, just a rebase left

Done.

h7x4 commented:
Cool! Thank you for the contribution! Welcome to nixpkgs :tada:
